File: clients/sdk/src/metadata.test.ts
import assert from 'node:assert/strict';
import { createHash } from 'node:crypto';
import { readFileSync } from 'node:fs';
import { test } from 'node:test';
import { fileURLToPath } from 'node:url';

import { PaymentMetadata } from './generated/blockchain/paymsg/v1/payment_metadata.ts';
import {
  METADATA_HASH_BYTES,
  METADATA_SALT_BYTES,
  metadataHash,
  newPaymentMetadata,
  verifyMetadata,
} from './metadata.ts';
import { payment } from './signing.ts';

/** The file x/paymsg/types reads too. See the vector test below. */
const vectors = JSON.parse(
  readFileSync(fileURLToPath(new URL('../../../testdata/vectors/confidentiality.json', import.meta.url)), 'utf8'),
) as {
  payment_metadata: Array<{
    name: string;
    salt_hex: string;
    purpose_code: string;
    remittance_information: string;
    wire_hex: string;
    hash_hex: string;
  }>;
};

function fromHex(hex: string): Uint8Array {
  const out = new Uint8Array(hex.length / 2);
  for (let i = 0; i < out.length; i += 1) out[i] = parseInt(hex.slice(i * 2, i * 2 + 2), 16);
  return out;
}

function toHex(bytes: Uint8Array): string {
  return Array.from(bytes, (b) => b.toString(16).padStart(2, '0')).join('');
}

test('a payload round-trips and its hash verifies', async () => {
  const payload = newPaymentMetadata('SALA', 'March salary, employee 4417');
  assert.equal(payload.salt.length, METADATA_SALT_BYTES);

  const hash = await metadataHash(payload);
  assert.equal(hash.length, METADATA_HASH_BYTES);
  assert.equal(await verifyMetadata(payload, hash), true);

  // The same payload hashes the same way every time, which is the whole basis
  // of the check.
  assert.deepEqual(await metadataHash(payload), hash);
});

test('an altered payload fails against the recorded hash', async () => {
  const payload = newPaymentMetadata('SALA', 'March salary, employee 4417');
  const hash = await metadataHash(payload);

  assert.equal(
    await verifyMetadata({ ...payload, remittanceInformation: 'March salary, employee 4418' }, hash),
    false,
  );
  assert.equal(await verifyMetadata({ ...payload, purposeCode: 'SUPP' }, hash), false);
  assert.equal(await verifyMetadata({ ...payload, remittanceInformation: '' }, hash), false);

  // A different salt is a different payload even when every readable field
  // matches, which stops one payment's payload being offered as another's.
  const other = newPaymentMetadata('SALA', 'March salary, employee 4417');
  assert.equal(await verifyMetadata({ ...payload, salt: other.salt }, hash), false);
});

test('two payments carrying the same detail do not hash alike', async () => {
  const first = await metadataHash(newPaymentMetadata('SALA', 'March salary'));
  const second = await metadataHash(newPaymentMetadata('SALA', 'March salary'));
  assert.notDeepEqual(first, second);
});

test('a payload with no salt is refused rather than hashed', async () => {
  await assert.rejects(
    () => metadataHash({ salt: new Uint8Array(4), purposeCode: 'SALA', remittanceInformation: '' }),
    /salt must be 32 bytes/,
  );
});

test('a recorded hash of the wrong length never verifies', async () => {
  const payload = newPaymentMetadata('SALA', 'March salary');
  const hash = await metadataHash(payload);
  assert.equal(await verifyMetadata(payload, hash.slice(0, 16)), false);
});

// The hash the browser computes has to be the hash the chain recorded, and the
// protobuf encoding underneath it has to be the same encoding.
//
// Both are read from testdata/vectors/confidentiality.json, which the Go suite
// reads too. This test used to rebuild the expected wire bytes by hand while
// the Go side pinned a hex constant: the two agreed, but neither could make the
// other fail, so the drift they were meant to catch would have gone through
// both of them. One file makes the contract real.
//
// The wire bytes are asserted as well as the digest. A digest alone says the
// two disagree; the encoding says where — a reordered field, a default that
// stopped being omitted.
test('payment metadata hashing matches the shared cross-language vectors', async () => {
  assert.ok(vectors.payment_metadata.length > 0, 'the vectors are empty, so this test would pass vacuously');

  for (const v of vectors.payment_metadata) {
    const payload = {
      salt: fromHex(v.salt_hex),
      purposeCode: v.purpose_code,
      remittanceInformation: v.remittance_information,
    };

    const wire = Uint8Array.from(PaymentMetadata.encode(payload).finish());
    assert.equal(toHex(wire), v.wire_hex, `${v.name}: the protobuf encoding of the payload has changed`);
    assert.equal(
      toHex(new Uint8Array(createHash('sha256').update(wire).digest())),
      v.hash_hex,
      `${v.name}: the digest does not match the encoding beside it`,
    );

    assert.equal(toHex(await metadataHash(payload)), v.hash_hex, v.name);
    assert.equal(await verifyMetadata(payload, fromHex(v.hash_hex)), true, v.name);
  }
});

test('a payment refuses to carry the hash and the plaintext together', async () => {
  const hash = await metadataHash(newPaymentMetadata('SALA', 'March salary'));

  assert.throws(
    () =>
      payment({
        debtor: 'yml1debtor',
        endToEndId: 'E2E-1',
        instructingParticipant: 'yml1one',
        instructedParticipant: 'yml1two',
        creditor: 'yml1creditor',
        denom: 'uyml',
        amount: '1000',
        purposeCode: 'SALA',
        metadataHash: hash,
      }),
    /never both/,
  );
});

test('a payment refuses a settlement jurisdiction that is not alpha-2', () => {
  const base = {
    debtor: 'yml1debtor',
    endToEndId: 'E2E-2',
    instructingParticipant: 'yml1one',
    instructedParticipant: 'yml1two',
    creditor: 'yml1creditor',
    denom: 'uyml',
    amount: '1000',
  };

  const jurisdictionOf = (msg: ReturnType<typeof payment>): string =>
    (msg.value as { settlementJurisdiction: string }).settlementJurisdiction;

  assert.equal(jurisdictionOf(payment({ ...base, settlementJurisdiction: 'NG' })), 'NG');
  // Optional today, so an omitted one is not an error here either.
  assert.equal(jurisdictionOf(payment(base)), '');

  for (const bad of ['nga', 'ng', 'NGA', 'N', 'N1']) {
    assert.throws(() => payment({ ...base, settlementJurisdiction: bad }), /alpha-2/, `accepted ${bad}`);
  }
});

File: x/paymsg/types/metadata_test.go
package types_test

import (
	"bytes"
	"encoding/hex"
	"testing"

	"github.com/stretchr/testify/require"

	"yamale/blockchain/x/paymsg/types"
)

// A payload built off-chain hashes to something the chain can record, and the
// party holding that payload can show it is the one recorded.
func TestPaymentMetadataRoundTrips(t *testing.T) {
	payload, err := types.NewPaymentMetadata("SALA", "March salary, employee 4417")
	require.NoError(t, err)
	require.Len(t, payload.Salt, types.MetadataSaltLength)

	hash, err := payload.Hash()
	require.NoError(t, err)
	require.Len(t, hash, types.MetadataHashLength)

	require.NoError(t, types.VerifyMetadata(payload, hash))

	// The same payload hashes the same way every time, which is the whole basis
	// of the check: a hash that depended on anything but the payload would
	// verify for the party that computed it and for nobody else.
	again, err := payload.Hash()
	require.NoError(t, err)
	require.True(t, bytes.Equal(hash, again))
}

// Altering any part of the payload breaks the proof. This is the property the
// off-chain store is worth nothing without: whoever holds the payload must not
// be able to produce a different one and call it the record.
func TestAlteredPaymentMetadataFailsVerification(t *testing.T) {
	payload, err := types.NewPaymentMetadata("SALA", "March salary, employee 4417")
	require.NoError(t, err)

	hash, err := payload.Hash()
	require.NoError(t, err)

	altered := []struct {
		name   string
		mutate func(types.PaymentMetadata) types.PaymentMetadata
	}{
		{
			name: "remittance information rewritten",
			mutate: func(m types.PaymentMetadata) types.PaymentMetadata {
				m.RemittanceInformation = "March salary, employee 4418"
				return m
			},
		},
		{
			name: "purpose code swapped",
			mutate: func(m types.PaymentMetadata) types.PaymentMetadata {
				m.PurposeCode = "SUPP"
				return m
			},
		},
		{
			name: "remittance information emptied",
			mutate: func(m types.PaymentMetadata) types.PaymentMetadata {
				m.RemittanceInformation = ""
				return m
			},
		},
		{
			// A different salt is a different payload even when every readable
			// field matches, which is what stops one payment's payload being
			// offered as another's.
			name: "salt replaced",
			mutate: func(m types.PaymentMetadata) types.PaymentMetadata {
				other, err := types.NewPaymentMetadata(m.PurposeCode, m.RemittanceInformation)
				require.NoError(t, err)
				m.Salt = other.Salt
				return m
			},
		},
	}

	for _, tc := range altered {
		t.Run(tc.name, func(t *testing.T) {
			err := types.VerifyMetadata(tc.mutate(payload), hash)
			require.Error(t, err)
			require.ErrorIs(t, err, types.ErrInvalidMetadata)
		})
	}
}

// Two payments carrying identical detail must not produce identical hashes, or
// the ledger publishes that they are the same payment detail without anybody
// ever decrypting anything.
func TestPaymentMetadataSaltsDifferPerPayment(t *testing.T) {
	first, err := types.NewPaymentMetadata("SALA", "March salary")
	require.NoError(t, err)
	second, err := types.NewPaymentMetadata("SALA", "March salary")
	require.NoError(t, err)

	firstHash, err := first.Hash()
	require.NoError(t, err)
	secondHash, err := second.Hash()
	require.NoError(t, err)

	require.False(t, bytes.Equal(firstHash, secondHash))
}

// The chain and the client SDK must produce the same digest for the same
// payload, or a payload that verifies in a browser fails against the block it
// was recorded in.
//
// Both sides read testdata/vectors/confidentiality.json, and that is the whole
// point. This test used to pin a hex constant here while the TypeScript side
// rebuilt the wire bytes by hand: the two agreed, but neither could ever make
// the other fail, so the drift they were meant to catch would have gone through
// both. Reading one file means an implementation that moves goes red against a
// fixture it cannot quietly rewrite.
//
// The wire encoding is asserted as well as the digest. A digest alone says the
// two disagree; the encoding says where — a reordered field, a default that
// stopped being omitted — which is the difference between a fix and an
// afternoon.
func TestPaymentMetadataMatchesTheSharedVectors(t *testing.T) {
	for _, v := range loadVectors(t).Metadata {
		t.Run(v.Name, func(t *testing.T) {
			payload := types.PaymentMetadata{
				Salt:                  mustHex(t, v.SaltHex),
				PurposeCode:           v.PurposeCode,
				RemittanceInformation: v.RemittanceInformation,
			}

			wire, err := payload.Marshal()
			require.NoError(t, err)
			require.Equal(t, v.WireHex, hex.EncodeToString(wire),
				"the protobuf encoding of the payload has changed")

			hash, err := payload.Hash()
			require.NoError(t, err)
			require.Equal(t, v.HashHex, hex.EncodeToString(hash))

			require.NoError(t, types.VerifyMetadata(payload, mustHex(t, v.HashHex)))
		})
	}
}

// A recorded value that is not a SHA-256 digest is reported as that, not as a
// payload mismatch. bytes.Equal would refuse it either way, so the check only
// earns its place by saying which side is wrong: a party whose store returned a
// truncated hash would otherwise spend the afternoon looking for a tampered
// remittance line that was never tampered with.
func TestVerifyMetadataDistinguishesAMalformedRecordedHash(t *testing.T) {
	payload, err := types.NewPaymentMetadata("SALA", "March salary")
	require.NoError(t, err)
	hash, err := payload.Hash()
	require.NoError(t, err)

	err = types.VerifyMetadata(payload, hash[:16])
	require.ErrorIs(t, err, types.ErrInvalidMetadata)
	require.Contains(t, err.Error(), "recorded hash must be")

	// And the payload-mismatch path still reports the other thing.
	tampered := payload
	tampered.PurposeCode = "SUPP"
	err = types.VerifyMetadata(tampered, hash)
	require.ErrorIs(t, err, types.ErrInvalidMetadata)
	require.Contains(t, err.Error(), "does not hash to the value recorded")
}

// The ISO 20022 limits follow the detail off-chain. A payload the receiving
// system cannot hold is worth failing on while the sender is still there to be
// told, rather than at delivery when they are not.
func TestPaymentMetadataEnforcesISOLimits(t *testing.T) {
	payload, err := types.NewPaymentMetadata("SALARY", "fine")
	require.NoError(t, err)
	_, err = payload.Hash()
	require.ErrorIs(t, err, types.ErrInvalidMetadata)
	require.Contains(t, err.Error(), "purpose_code")

	long := make([]byte, types.MaxRemittanceLength+1)
	for i := range long {
		long[i] = 'x'
	}
	payload, err = types.NewPaymentMetadata("SALA", string(long))
	require.NoError(t, err)
	_, err = payload.Hash()
	require.ErrorIs(t, err, types.ErrInvalidMetadata)
	require.Contains(t, err.Error(), "remittance_information")
}

// A payload with no salt is refused rather than hashed, because a hash of a
// four-character purpose code with nothing else in it is a lookup table.
func TestPaymentMetadataRefusesUnsaltedPayload(t *testing.T) {
	_, err := types.PaymentMetadata{PurposeCode: "SALA"}.Hash()
	require.ErrorIs(t, err, types.ErrInvalidMetadata)

	short := types.PaymentMetadata{Salt: []byte("too short"), PurposeCode: "SALA"}
	_, err = short.Hash()
	require.ErrorIs(t, err, types.ErrInvalidMetadata)
}

func TestValidateSettlementJurisdiction(t *testing.T) {
	testCases := []struct {
		name     string
		country  string
		required bool
		valid    bool
	}{
		{name: "NG", country: "NG", required: false, valid: true},
		{name: "NG when required", country: "NG", required: true, valid: true},
		{name: "GH", country: "GH", required: true, valid: true},

		// Lowercase matches no authority in the registry, so it settles with
		// nobody while reading on the record as though somebody was named.
		{name: "nga", country: "nga", required: false, valid: false},
		{name: "ng", country: "ng", required: false, valid: false},

		// alpha-3 is the other ISO 3166-1 list, and the one a caller reaches
		// for by mistake.
		{name: "NGA", country: "NGA", required: false, valid: false},

		{name: "N", country: "N", required: false, valid: false},
		{name: "N1", country: "N1", required: false, valid: false},
		{name: "N ", country: "N ", required: false, valid: false},

		// Two bytes, one rune: the length check alone would let this through.
		{name: "non-ascii", country: "Ñ", required: false, valid: false},

		{name: "empty when optional", country: "", required: false, valid: true},
		{name: "empty when required", country: "", required: true, valid: false},
	}

	for _, tc := range testCases {
		t.Run(tc.name, func(t *testing.T) {
			err := types.ValidateSettlementJurisdiction(tc.country, tc.required)
			if tc.valid {
				require.NoError(t, err)
				return
			}
			require.Error(t, err)
			require.ErrorIs(t, err, types.ErrInvalidSettlementJurisdiction)
		})
	}
}

func TestValidateConfidentiality(t *testing.T) {
	payload, err := types.NewPaymentMetadata("SALA", "March salary")
	require.NoError(t, err)
	hash, err := payload.Hash()
	require.NoError(t, err)

	t.Run("hash alone is accepted", func(t *testing.T) {
		require.NoError(t, types.ValidateConfidentiality(nil, nil, hash, "", ""))
	})

	t.Run("plaintext alone is accepted", func(t *testing.T) {
		require.NoError(t, types.ValidateConfidentiality(nil, nil, nil, "SALA", "March salary"))
	})

	t.Run("hash beside the plaintext it replaces is refused", func(t *testing.T) {
		err := types.ValidateConfidentiality(nil, nil, hash, "SALA", "")
		require.ErrorIs(t, err, types.ErrInvalidMetadata)

		err = types.ValidateConfidentiality(nil, nil, hash, "", "March salary")
		require.ErrorIs(t, err, types.ErrInvalidMetadata)
	})

	t.Run("a hash of the wrong length is refused", func(t *testing.T) {
		err := types.ValidateConfidentiality(nil, nil, []byte("not a sha256 digest"), "", "")
		require.ErrorIs(t, err, types.ErrInvalidMetadata)
	})

	t.Run("commitments are refused while unverified", func(t *testing.T) {
		err := types.ValidateConfidentiality([]byte{0x01}, nil, nil, "", "")
		require.ErrorIs(t, err, types.ErrConfidentialAmountUnavailable)

		err = types.ValidateConfidentiality(nil, []byte{0x02}, nil, "", "")
		require.ErrorIs(t, err, types.ErrConfidentialAmountUnavailable)
	})
}
